visions/accessors.md

```diff
@@ -0,0 +1,348 @@
+
+### Ownership of the containing value
+
+All of the above accessors include a reading variant (`get`, `read`, `borrow`) and a writing variant (`set`, `modify`, `mutate`).  By default, the reading variant is not considered to be a mutation of the enclosing value.  The writing variant conversely is considered to be a write access of the enclosing value and the usual Swift exclusivity rules apply to limit the scope of such access.
+
+However, a reading accessor can be explicitly marked as `mutating`.  This relaxes checks on the accessor implementation so that it can mutate the containing value, and also causes the caller to treat this as a write access requiring exclusivity enforcement.  This is commonly used for caching or bookkeeping:
+
+```swift
+struct Foo {
+  private var accessCount: Int
+  private var _cachedBaz: Baz?
+  var baz: Baz {
+     mutating get {
+       accessCount += 1 // OK because of `mutating`
+       if _cachedBaz == nil {
+         _cachedBaz = computeBazPropertyValue() // OK because of `mutating`
+       }
+       return _cachedBaz
+     }
+  }
+}
+```
+
+Conversely, writing accessors can be marked as `nonmutating`, though this is not generally useful in practice as it would require that the accessor not in fact mutate the containing value.  This could only be useful for modeling non-property operations as setters.
+
+Accessors can also be `consuming`.  This indicates that the containing object must end its lifetime once this access is complete.  This is useful for representing certain types of object transformations:
+
+```swift
+struct Values {
+  var dropOldest: Values {
+    consuming get {
+       let newValues = ... everything but the oldest ...
+       return newValues
+       // Current value will end it's lifetime after this returns
+    }
+  }
+}
+
+let v = Values()
+let v2 = v.dropOldest
+// `v` is no longer alive here
+```
+
+Note that `consuming` makes no sense for setters — there is no point to changing a property on a value and then immediately ending the lifetime of that value.  Similarly, `consuming` makes no sense for `borrow` or `unsafeAddress` accessors — those require that the containing value survive for the duration of the returned borrow access or address, so it does not make sense to explicitly terminate the value lifetime immediately.
+
+This only leaves `consuming get` and `consuming read` as meaningful combinations.  A `consuming get` can be used for operations such as the `dropOldest` example above that model a transformation by creating a new value and terminating the old one.  The `consuming read` variant can be used similarly.  Unlike `borrow`, the `read` operation coroutine structure forces the containing value to live for a certain period of time, the value is consumed only at the end of that coroutine.
```

Comment:
In a `consuming get`, I imagine that the object is `deinit`ed at some point before the accessor returns. In a `consuming read`, is it `deinit`ed at the return of the top half or the return of the bottom half?

Reply:
I'm not entirely certain whether technically the owning value is destroyed before or after the accessor returns.  (@jckarter would certainly know.)  In the coroutine case, we could not invoke the bottom half if the value were already destroyed, so it would have to be destroyed at the end of the bottom half.